### ISSUE TITLE
iOS 10 handleInputModeList issue

### ISSUE THREAD

**Issue by pirrate** [archagon/tasty-imitation-keyboard#107]
I am trying to implement handleInputModeList that was added in iOS10. This is my code
  case Key.KeyType.keyboardChange:
                           if #available(iOSApplicationExtension 10.0, *) {
                                keyView.addTarget(self, action: #selector(self.handleInputModeList(from:with:)), for:UIControlEvents.allTouchEvents)
                            } else {
        
                                keyView.addTarget(self, action: #selector(KeyboardViewController.advanceTapped(_:)), for: .touchUpInside)
                            }

But it doesn`t work. I tried to create new button and add same target there and it worked normally. But it doesn`t work with keyViews. Any idea?

**Comment by pirrate** [comment]
@archagon @AaronFW any idea how to handle it?

**Comment by AaronFW** [comment]
I haven't had much chance to try using it.

**Comment by AaronFW** [comment]
Well, I think it is because handleInputModeList(from:with) is a method of UIInputViewController but  keyView is KeyboardKey Class and KeyboardKey Class is UIControl (which inherits from UIView). 

Typically when handleInputModeList is used, it is used directly in the UIInputControlView class declaration. For example, one project I looked at had
` self.nextKeyboardButton.addTarget(self, action: #selector(handleInputModeList(from:with:)), for: .allTouchEvents)`
In which "self" was the `class KeyboardControllerView: UIInputControllerView `

**Comment by pirrate** [comment]
@AaronFW I think the main reason is because @archagon rewriting events on forwardingview and handleInputModeList(from:with:) is something that it doesn`t understand. Normally if i add extra UIButton to forwardingview and add same target it works without any problem. Now i don`t know how to figure it out with this architecture.

**Comment by mromanuk** [comment]
@pirrate you are right, I needed to change the way ForwardingView and KeyboardKey track touch.  

That allow me to do this in `keyboardViewController.setupKeys()`:

```swift
keyView.addTarget(self, action: #selector(handleInputModeList(from:with:)), for: .allTouchEvents)
keyView.originalViewTouch = true
```

the `originalViewTouch` is a flag to use the *native* UIControl tracking capability instead of the Archagon keyboard tracking in `ForwardingView.hitTest`

**Comment by ifanchu** [comment]
@mromanuk do you mind shedding some light on how to implement this? Some code snippets maybe? Appreciated.

**Comment by ifanchu** [comment]
I think I figure it out, thanks for the tip @mromanuk

**Comment by vishal07malvi** [comment]
@ifanchu @mromanuk @archagon @AaronFW could you please let me know this issue is fixed or not? because I also having issue with 
`     keyView.addTarget(self, action: #selector(self.handleInputModeList(from:with:)), for:UIControlEvents.allTouchEvents)
`
handleInputModeList, Could you please let me know what should I do with this?

**Comment by kamalupasena** [comment]
@mromanuk can you share a code sample showing how you do it ?

**Comment by kamalupasena** [comment]
@ifanchu Can you share a code sample ?

**Comment by aalenliang** [comment]
Here is my full solution to this issue, don't know if it's the best approach. thanks for @mromanuk sharing the tips.

in **KeyboardKey.swift**, add `originalViewTouch` property:
```
var originalViewTouch = false
```

in **KeyboardViewController.swift**, `setupKeys` function:
```
case Key.KeyType.keyboardChange:
    keyView.addTarget(self,
                      action: #selector(handleInputModeList(from:with:)),
                      for: .allTouchEvents)
    keyView.originalViewTouch = true
```
and delete `advanceTapped` method.

in **ForwardingView.swift**, `hitTest` function:
replace 
```
else {
    return (self.bounds.contains(point) ? self : nil)
}
```
with
```
if !self.bounds.contains(point) {
    return nil
}
if let view = self.findNearestView(point),
    let keyView = view as? KeyboardKey,
    keyView.originalViewTouch {
    return keyView
}
return self
```
